### PR TITLE
improve openapi v3 schema for appcatalog CRD

### PR DESCRIPTION
**Pull request by puja108**
towards https://github.com/giantswarm/giantswarm/issues/8169

### CODE REVIEW

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 1]
```diff
@@ -47,7 +47,6 @@ spec:
                       type: string
                     resourceVersion:
                       type: string
-                  required: ["name", "namespace"]
```
Comment (puja108) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: config is not required

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 2]
```diff
@@ -57,19 +56,21 @@ spec:
...
```
Comment (puja108) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: secret is also not required

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 3]
```diff
@@ -57,19 +56,21 @@ spec:
...
                 type:
                   type: string
+				  enum: ["helm"]
+				  default: helm
                 URL:
                   type: string
                   format: uri 
               required: ["type", "URL"]
-          required: ["title", "description", "storage"]
+          required: ["title", "storage"]
```
Comment (puja108) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: these are the only 2 needed to work

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 4]
```diff
@@ -57,19 +56,21 @@ spec:
...
-                  required: ["name", "namespace"]
             logoURL:
               type: string
+			  format: uri
             storage:
               type: object 
               properties:
                 type:
                   type: string
+				  enum: ["helm"]
```
Comment (puja108) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: here we can add other types in the future

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 5]
```diff
@@ -57,19 +56,21 @@ spec:
...
+				  default: helm
                 URL:
                   type: string
                   format: uri 
               required: ["type", "URL"]
-          required: ["title", "description", "storage"]
+          required: ["title", "storage"]
 `
```
Comment (puja108) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: I saw that in https://kubernetes.io/blog/2019/06/20/crd-structural-schema/ the also add `required: ["spec"]`, do you think that is needed? isn't it kinda implied by having required on the lower level? and if latter is true then do we need to make `storage` required if the sub keys are already?

**Comment by puja108** [comment]
Could we maybe add additional validation as e.g. some of our fields are strings but do not allow for spaces or special characters?

**Comment by puja108** [comment]
Same I also saw in https://kubernetes.io/blog/2019/06/20/crd-structural-schema/ that they often use `minLength: 1` to not get empty fields, especially for required fields that might make sense. WDYT?

File: pkg/apis/application/v1alpha1/app_catalog_types.go (hunk 1 above)
Comment (sslavic) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: config is not required, but once it is defined, name and namespace for it are required

File: pkg/apis/application/v1alpha1/app_catalog_types.go (hunk 2 above)
Comment (sslavic) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: secret is not required, but once it is defined, name and namespace for it are required

File: pkg/apis/application/v1alpha1/app_catalog_types.go (hunk 4 above)
Comment (sslavic) [reply pkg/apis/application/v1alpha1/app_catalog_types.go]: formatting/indentation issue

**Comment by sslavic** [comment]
Thanks for working on improving the AppCatalog CRD!

Would be great to test this apiextensions branch in an app-operator branch (add override for apiextensions on Gopkg.toml), before merging this PR to master.

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 6]
```diff
@@ -60,16 +60,20 @@ spec:
                   required: ["name", "namespace"]
             logoURL:
               type: string
+              format: uri
             storage:
               type: object 
               properties:
                 type:
                   type: string
+                  enum: ["helm"]
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_catalog_types.go]: Nice!

**Comment by tfussell** [comment]
Want a hand translating these improvements into `kubebuilder` validation comments?

**Comment by xh3b4sd** [comment]
Can we merge or close this PR?

**Comment by puja108** [comment]
This was handed over to @giantswarm/team-batman-engineers (specifically 
@pipo02mix IIRC, but he now also moved teams).

**Comment by xh3b4sd** [comment]
If anyone wants to pick this up feel free to reopen.

**Pull request closed without merging**